python/lsst/resources/_resourceHandles/_s3ResourceHandle.py

```diff
@@ -0,0 +1,260 @@
+        # written to.
+        if (
+            (self.tell() - (self._last_flush_position or 0)) < 5242880
+            and not self._closed == CloseStatus.CLOSING
+            and not self._warned
+        ):
+            warnings.warn("S3 does not support flushing objects less than 5Mb, skipping")
+            self._warned = True
+            return
+        # nothing to write, don't create an empty upload
+        if self.tell() == 0:
+            return
+        with time_this(
+            self._log,
+            msg="Upload multipart %d to %s",
+            args=(
+                self._partNo,
+                self,
+            ),
+        ):
+            response = self._client.upload_part(
+                Body=self._buffer.getvalue(),
+                Bucket=self._bucket,
+                Key=self._key,
+                UploadId=self._multiPartUpload["UploadId"],
+                PartNumber=self._partNo,
+            )
+        self._parts.append({"PartNumber": self._partNo, "ETag": response["ETag"]})
+        self._partNo += 1
+        self._last_flush_position = self._buffer.tell() + (self._last_flush_position or 0)
+        self._buffer = BytesIO()
+
+    @property
+    def isatty(self) -> bool:
+        return False
+
+    def readable(self) -> bool:
+        return self._readable
+
+    def readline(self, size: int = -1) -> bytes:
+        raise OSError("S3 Does not support line by line reads")
+
+    def readlines(self, hint: int = -1) -> Iterable[bytes]:
+        self.seek(0)
+        return self.read().split(self._newline)
+
+    def seek(self, offset: int, whence: int = SEEK_SET) -> int:
+        if self.writable():
+            if self._last_flush_position is not None:
+                if whence == SEEK_SET:
+                    offset -= self._last_flush_position
+                    if offset < 0:
+                        raise OSError("S3 ResourceHandle can not seek prior to already flushed positions")
+                if whence == SEEK_CUR:
+                    if (self.tell() - self._last_flush_position) < 0:
+                        raise OSError("S3 ResourceHandle can not seek prior to already flushed positions")
+                if whence == SEEK_END:
+                    raise OSError("S3 ResourceHandle can not seek referencing the end of the resource")
+            self._buffer.seek(offset, whence)
+            self._position = self._buffer.tell()
+        else:
+            if whence == SEEK_SET:
+                self._position = offset
+            elif whence == SEEK_CUR:
+                self._position += offset
+            elif whence == SEEK_END:
+                offset = abs(offset)
+                self._position -= offset
+        return self._position
+
+    def seekable(self) -> bool:
+        return True
+
+    def truncate(self, size: Optional[int] = None) -> int:
+        if self.writable():
+            self._buffer.truncate(size)
```

Comment:
What if you've flushed before you call truncate? If I call `truncate(0)` I expect to have set the file size to 0 so I can start again but it seems in this code if I call `flush(); truncate(0)` the truncate won't reset the file to 0 bytes because the buffer has been cleared by the flush and sent to S3 already. Am I missing something?

Reply:
So I normally would have expected the normal course of this to be `flush(); truncate(0); flush()` if you wanted to clear out the file. That said thinking about this, it may not like that with filesize uploads. This might require more thought.